prog/dftb+/lib_dftb/extfields.F90

```diff
@@ -0,0 +1,185 @@
+!--------------------------------------------------------------------------------------------------!
+!  DFTB+: general package for performing fast atomistic simulations                                !
+!  Copyright (C) 2006 - 2021  DFTB+ developers group                                               !
+!                                                                                                  !
+!  See the LICENSE file for terms of usage and distribution.                                       !
+!--------------------------------------------------------------------------------------------------!
+
+!> External fields
+module dftbp_extfields
+  use dftbp_accuracy, only : dp, lc
+  use dftbp_potentials, only : TPotentials
+  use dftbp_periodic, only : TNeighbourList
+  use dftbp_message, only : error
+  implicit none
+
+  public :: TEField, TElecField_inp, setUpExternalField
+
+  !> External field data structure
+  type :: TEField
+
+    ! homogeneous electric field
+
+    !> field strength
+    real(dp), allocatable :: EFieldStrength
+
+    !> field direction
+    real(dp) :: EfieldVector(3)
+
+    !> Field is time dependent
+    logical :: isTDEfield
+
+    !> angular frequency
+    real(dp) :: EfieldOmega
+
+    !> phase of field at step 0
+    integer :: EfieldPhase
+
+    ! Derived properties from E field
+
+    !> Instantaneous electric field
+    real(dp) :: EField(3)
+
+    !> Instantaneous magnitude of electric field
+    real(dp) :: absEField
+
+    ! potentials at atomic sites
+
+    !> Atoms at which there is a external potential
+    integer, allocatable :: atomicSites(:)
+
+    !> Potential at the atomic sites
+    real(dp), allocatable :: atomicPotential(:)
+
+    !> is this an onsite (net charge related, so diagonal elements of hamiltonian) or not
+    !> (gross/Mulliken charge related, so incluing off diagonal coupling)
+    logical, allocatable :: atomicOnSite(:)
+
+  end type TEField
+
+
+  !> External homogeneous electric field
+  type TElecField_inp
+
+    !> time dependent field in MD
+    logical :: isTDEfield = .false.
+
+    !> strength
+    real(dp) :: EFieldStrength = 0.0_dp
+
+    !> direction
+    real(dp) :: EfieldVector(3)
+
+    !> frequency of time dependent field
+    real(dp) :: EfieldOmega
+
+    !> relative phase of field
+    integer :: EfieldPhase = 0
+
+  end type TElecField_inp
+
+contains
+
+  !> Sets up external electric or other fields
+  subroutine setUpExternalField(eField, tPeriodic, neighbourList, nNeighbourSK, iCellVec,&
+      & img2CentCell, cellVec, deltaT, iGeoStep, coord0Fold, coord, potential)
```

Comment:
Maybe `addExternalField` is a better name?

Reply:
Done